File: claudeApi.js
const Anthropic = require("@anthropic-ai/sdk");
const { performance } = require("perf_hooks");
const { searchNews } = require('./newsSearch.js');
const moment = require('moment');
const config = require('./config.js');
const { countTokens, calculateCost } = require('./utils.js');
const ConversationManager = require('./conversationManager.js');

const anthropic = new Anthropic({ apiKey: process.env.CLAUDE_API_KEY });
const conversationManager = new ConversationManager();

async function determineSearchNeed(userHistory, currentDateTime) {
  console.log("Sending initial request to Claude Sonnet");
  
  // Ensure the first message has a "user" role
  const messages = userHistory.length > 0 && userHistory[0].role === "user" 
    ? userHistory 
    : [{ role: "user", content: "Hello" }, ...userHistory];

  const initialResponse = await anthropic.messages.create({
    model: config.INITIAL_MODEL,
    max_tokens: config.INITIAL_MAX_TOKENS,
    system: `You are an AI assistant. The current date and time is ${currentDateTime}. Determine if the user's query requires current or future information that would benefit from a web search. If so, respond with 'SEARCH:query', where 'query' is a concise search term. Otherwise, respond with 'NO_SEARCH'.`,
    messages: messages,
  });
  return initialResponse.content[0].text;
}

async function performNewsSearch(searchQuery) {
  console.log(`Performing news search for: ${searchQuery}`);
  try {
    const newsResults = await searchNews(searchQuery);
    return newsResults.map((result, index) => 
      `[${index + 1}] ${result.title}\nSource: ${result.source}\nPublished: ${result.publishedAt}\nURL: ${result.link}\nSummary: ${result.snippet || 'No summary available.'}\n`
    ).join('\n').trim();
  } catch (searchError) {
    console.error("News search failed:", searchError);
    throw new Error(`News search failed for query: "${searchQuery}"`);
  }
}

async function askClaude(userId, message) {
  const startTime = performance.now();

  try {
    let userHistory = conversationManager.getUserHistory(userId, message);
    
    // Ensure there's at least one user message in the history
    if (userHistory.length === 0 || userHistory[0].role !== "user") {
      userHistory.unshift({ role: "user", content: message });
    }

    const initialInputTokens = countTokens(JSON.stringify(userHistory));
    console.log(`Initial input tokens: ${initialInputTokens}`);

    const currentDateTime = moment().format('MMMM D, YYYY HH:mm:ss');
    const initialResponse = await determineSearchNeed(userHistory, currentDateTime);

    let searchSummary = 'No web search performed';
    let webSearchPerformed = false;
    let webSearchTokens = 0;
    if (initialResponse.startsWith("SEARCH:")) {
      const searchQuery = initialResponse.slice(7).trim();
      const formattedResults = await performNewsSearch(searchQuery);
      userHistory[userHistory.length - 1].content += `\n\n\nWeb search results for "${searchQuery}":\n\n${formattedResults}`;
      searchSummary = `Search results for "${searchQuery}":\n${formattedResults}`;
      webSearchPerformed = true;
      webSearchTokens = countTokens(formattedResults);
    } else {
      searchSummary = `No web search performed. Initial response: ${initialResponse.trim()}`;
    }

    const finalInputTokens = countTokens(JSON.stringify(userHistory));
    console.log(`Final input tokens: ${finalInputTokens}`);

    const claudeResponse = await anthropic.messages.create({
      model: config.FINAL_MODEL,
      max_tokens: config.FINAL_MAX_TOKENS,
      system: config.SYSTEM_PROMPT(currentDateTime),
      messages: userHistory.map(msg => ({ ...msg, content: msg.content.trim() })),
    });

    if (!claudeResponse.content || claudeResponse.content.length === 0) {
      throw new Error("Empty response from Claude API");
    }

    const outputTokens = countTokens(claudeResponse.content[0].text);
    console.log(`Output tokens: ${outputTokens}`);

    const initialCost = calculateCost(initialInputTokens, 0);
    const webSearchCost = calculateCost(webSearchTokens, 0);
    const finalCost = calculateCost(finalInputTokens, outputTokens);
    const totalCost = initialCost + webSearchCost + finalCost;

    const finalResponse = claudeResponse.content[0].text.trim();
    conversationManager.updateUserHistory(userId, finalResponse);

    const endTime = performance.now();
    const processingTime = (endTime - startTime).toFixed(2);
    console.log(`Request processed in ${processingTime}ms`);

    return {
      text: finalResponse,
      initialInputTokens,
      webSearchTokens,
      finalInputTokens,
      outputTokens,
      initialCost,
      webSearchCost,
      finalCost,
      totalCost,
      processingTime,
      model: config.FINAL_MODEL,
      webSearchPerformed,
      searchSummary,
    };
  } catch (error) {
    console.error("Error calling Claude API:", error);
    return "Sorry, I encountered an error while processing your request. Please try again later.";
  }
}

module.exports = { askClaude, clearHistory: conversationManager.clearHistory };

File: claude_api.py
import anthropic
import os
from dotenv import load_dotenv
from config import Config
from datetime import datetime, timezone
from web_analyzer import LLMCrawler, CrawlResult
from search_service import google_search

load_dotenv()

client = anthropic.Anthropic(api_key=os.getenv('CLAUDE_API_KEY'))
conversation_history = {}
crawler = LLMCrawler(os.getenv('CLAUDE_API_KEY'))

async def ask_claude(user_id, message):
    print(f"\nReceived message from user {user_id}: {message}")
    if user_id not in conversation_history:
        conversation_history[user_id] = []
        print(f"Created new conversation history for user {user_id}")

    current_time = datetime.now(timezone.utc)
    time_info = f"Current date and time: {current_time.strftime('%Y-%m-%d %H:%M:%S')} UTC"

    search_decision_prompt = f"""{time_info}

    Determine if a web search is needed to answer this query: "{message}"
    Your task is to decide whether current, real-time information from the internet is necessary to provide an accurate and complete answer.
    Respond with ONLY 'Yes' or 'No'.
    'Yes' if a search is needed (e.g., for current events, rapidly changing information, or specific facts you're unsure about).
    'No' if you can confidently answer based on your existing knowledge."""

    try:
        search_decision_response = client.messages.create(
            model=Config.INITIAL_MODEL,
            max_tokens=Config.INITIAL_MAX_TOKENS,
            system="You are an AI assistant tasked with determining if a web search is needed.",
            messages=[
                {"role": "user", "content": search_decision_prompt}
            ]
        )
        
        search_decision = search_decision_response.content[0].text.strip().lower()
        search_needed = (search_decision == 'yes')
        print(f"Search needed: {search_needed}")
    except Exception as e:
        print(f"Error in search decision: {str(e)}. Defaulting to perform search.")
        search_needed = True

    if search_needed:
        try:
            search_results = google_search(message, num_results=3)
            crawl_results = []
            for url in search_results:
                result = crawler.crawl_and_analyze(url)
                crawl_results.append(result)
            
            combined_info = "\n\n".join([
                f"URL: {result.url}\nSummary: {result.summary}\nSentiment: {result.sentiment}\nKey Topics: {result.key_topics}"
                for result in crawl_results
            ])
            
            full_message = f"{time_info}\n\nHere is some relevant information I found:\n\n{combined_info}\n\nBased on this information, please answer the following user query: {message}"
            print(f"Full message: {full_message}")  
        except Exception as e:
            print(f"Error during web analysis: {str(e)}")
            full_message = f"{time_info}\n\nAn error occurred during the web analysis. User query: {message}"
    else:
        full_message = f"{time_info}\n\n{message}"

    conversation_history[user_id].append({"role": "user", "content": full_message})

    print(f"Sending request to Claude API for user {user_id}")
    try:
        response = client.messages.create(
            model=Config.FINAL_MODEL,
            max_tokens=Config.FINAL_MAX_TOKENS,
            messages=conversation_history[user_id]
        )

        conversation_history[user_id].append({"role": "assistant", "content": response.content[0].text})
        print(f"Received response from Claude API for user {user_id}")
        return response.content[0].text
    except Exception as e:
        print(f"Error calling Claude API: {str(e)}")
        return "I'm sorry, but I encountered an error while processing your request. Please try again later."

def clear_history(user_id):
    if user_id in conversation_history:
        conversation_history[user_id] = []
        print(f"Cleared conversation history for user {user_id}")
        return "Conversation history cleared."
    print(f"No conversation history found for user {user_id}")
    return "No conversation history found."

async def analyze_webpage(url: str) -> str:
    try:
        result = crawler.crawl_and_analyze(url)
        return (f"Analysis of {url}:\n"
                f"Summary: {result.summary}\n"
                f"Sentiment: {result.sentiment}\n"
                f"Key Topics: {result.key_topics}\n"
                f"Total Cost: ${result.cost_total:.6f}")
    except Exception as e:
        return f"Error analyzing webpage: {str(e)}"